[FAB-17786] Update upgrade_dbs peer command to drop state couchdb

wenjianqiao opened the pull request:
Signed-off-by: Wenjian Qiao <wenjianq@gmail.com

#### Type of change
- Improvement (improvement to code, performance, etc)

#### Description
The existing upgrade_dbs command does not automatically drop state couchdbs
and therefore a separate step is required to drop couchdbs, This PR updates
the command to automatically drop state couchdbs.

#### Related issues
https://jira.hyperledger.org/browse/FAB-17786

#### Release Note
TBD

srderson commented:
@angrbrd @smithbk Do the proto files have to be duplicated? Is it possible to share the files in the protos folder?

binhn commented:
@angrbrd how can we test this prior to merge?

angrbrd commented:
@srderson, I guess it is possible for us to pull in the .proto files directly from the /fabric/protos directory inside our .js files. But I guess the question is how do we do that when we package this as an npm package? Does it make sense to reference them from /fabric/protos when we are packaging the npm? Will that still work? My impression was that we have to have them inside the directory containing the Node project to package them. Let me know what you think.

angrbrd commented:
@binhn, if you are worried that it will break any of the existing code that is not a concern, as this submission does not interact with any of the existing code at all. As far as UT, I did start a UT file together with the submission that includes two tests covering current functionality. The test file is under /sdk/node/test/unit/chain-test.js.

binhn commented:
@angrbrd I know it doesn't break existing code, but want to make sure this code has test-cases that I can run prior to merge.  Thanks for the info.

binhn commented:
Regarding protos, we can either ask the developers to copy the protos from the peer over, or we copy as done during packaging. However, I think we need to introduce version somehow to make sure the nodejs sdk is at the same level as the peer.

binhn commented:
I scanned through the code and noticed that we also need to support SHA2, issue #908, for both 256 and 384

angrbrd commented:
@binhn , issue #908 should already be closed as the associated PR #1148 has been merged.

wenjianqiao committed:
[FAB-17786] Update upgrade_dbs peer command to drop state couchdb

The existing upgrade_dbs command does not automatically drop state couchdbs
and therefore a separate step is required to drop couchdbs, This PR updates
the command to automatically drop state couchdbs.

In addition, it checks upgrade eligibility before upgrade so that it will not
drop databases if it is already the expected format.

Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

wenjianqiao commented:
Hit ci flake FAB-16233

wenjianqiao commented:
/ci-run

cendhu commented on core/ledger/kvledger/upgrade_dbs.go:
```diff
@@ -24,29 +32,39 @@ func UpgradeDBs(rootFSPath string) error {
 	defer fileLock.Unlock()
```
Not related to this PR. We have this exact `fileLock` code in four places (updatedbs, rollback, reset, ledger initializing). If possible, it will be addressed as part of [FAB-17609](https://jira.hyperledger.org/browse/FAB-17609)

cendhu commented on core/ledger/kvledger/kv_ledger_provider.go:
```diff
@@ -498,6 +498,10 @@ func openIDStore(path string) (s *idStore, e error) {
 	return &idStore{db, path}, nil
 }
 
+func (s *idStore) getFormat() ([]byte, error) {
```
Let's see whether @manish-sethi like this tiny function 😄 
Anyway, in one of the comment, I have mentioned a way to remove this function.

manish-sethi replied:
I am not always against the tiny function. In this case, it does not create a problem. I feel this problematic when a logic that makes sense in a single flow is broken down in multiple tiny functions and eventually all those functions are just called one after another and further, in isolation, mostly, none of those functions justify their existence.

cendhu commented on core/ledger/kvledger/kv_ledger_provider_test.go:
```diff
@@ -504,7 +504,8 @@ func constructTestLedgerID(i int) string {
 
 func testConfig(t *testing.T) (conf *lgr.Config, cleanup func()) {
-	path, err := ioutil.TempDir("", "kvledger")
+	// use /tmp instead of os.TempDir so that we can mount dirs for couchdb
```
The default folders shared between mac osx and docker are 
- /Users/
- /Volumes/
- /private/  
- /tmp

The `os.TempDir` creates a temp folder in `/var/folders` and hence the mount would be denied. Given that the test passes in CI (linux environment), I assume that `/tmp` is also shared between the linux host and docker. 

We can update this comment to add more information as `instead of os.TempDir` would confuse us when we look at the code after a month.

cendhu commented on core/ledger/kvledger/upgrade_dbs.go:
```diff
@@ -24,29 +32,39 @@ func UpgradeDBs(rootFSPath string) error {
-	db := leveldbhelper.CreateDB(&leveldbhelper.Conf{DBPath: dbPath})
-	db.Open()
-	defer db.Close()
+func dropCouchAndLevelDBs(config *ledger.Config) error {
```
This function name is a bit confusing as we use either CouchDB or leveldb as statedb but not both. Though we use leveldb for all internal states, we don't explicitly name it leveldb instead call it `configHistoryDB`, `HistoryDB`, `BookkeeperDB`.

Further, the dropDBs() also try to drop the `stateleveldb`. Assume that couchDB is used as the stateDB. With that assumption, the code drops `statecouchdb` and also try to drop `stateleveldb`. Hence, It looks hacky.

As we can drop CouchDB during rollback/reset too, we can move this logic to `dropDBs()` itself and either drop whole statecouchdb or whole stateleveldb. However, this require changes to both rollback and reset code (need to pass `ledger.config` rather than the `ledger.config.rootfspath`). Hence, I am fine with doing it in a separate PR. When we implement snapshot processing, we can avoid dropping the whole DB and drop only the dbs associated with the channel.  At that time, we could infact pass a config struct to dropDBs with info such as statedb used, channels to be dropped (one or all), rootfspath, etc...). 

Wdyt @manish-sethi ?

wenjianqiao replied:
Prefer to use a separate PR for reset/rollback because they have user impact. As of dropDBs unnecessarily drops stateleveldb in the case of couchdb, it is true for other commands too. I can implement a new method instead of calling dropDBs.

manish-sethi replied:
IMO, better to leave `dropDBs` as is. Instead, rename drop `dropStateLevelDB` as `dropStateDB`. Then, within `dropStateDB`, you can check and delete one of the leveldb or couchdb.

cendhu replied:
The suggestion given by Manish seems better.

wenjianqiao replied:
Yes, agree with Manish's suggestion, but it means that we will drop statecouchdb for all reset/rollback/upgrade commands. Meanwhile, I try to limit the scope of this PR. Maybe we can use this PR to update dropDBs (and dropStateDB) for all 3 commands and use a separate PR to update integration tests. What do you think?

cendhu replied:
Yes, we can limit the scope of this PR as it simplifies the reviewer's task too.

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb_test_export.go:
```diff
@@ -24,15 +24,17 @@ func StartCouchDB(t *testing.T, binds []string) (addr string, stopCouchDBFunc fu
 
 // DeleteApplicationDBs deletes all the databases other than fabric internal database
 func DeleteApplicationDBs(t testing.TB, config *ledger.CouchDBConfig) {
```
Given that `DropApplicationDBs()` is exported from the `couchdbutil.go`, I don't think we need `DeleteApplicationDBs()` in the `test_export.go`. Wherever we have called `DeleteApplicationDBs()`, now we can call `DropApplicationDBs()` and assert on the returned error..

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil.go:
```diff
@@ -297,3 +298,28 @@ func escapeUpperCase(dbName string) string {
+		if err != nil {
+			logger.Errorf("Error dropping CouchDB database %s", dbName)
+			return err
+		}
```
Earlier, we called `dropDB()` present in the `test_export.go`  instead of these lines. Now, it would look wrong to keep `dropDB()` in the `test_export.go` as it is used only by the local test file. Hence, I would suggest replacing these lines with `dropDB()` and move the `dropDB()` to this file & remove assertions.

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil_test.go:
```diff
@@ -184,3 +186,35 @@ func TestConstructedNamespaceDBName(t *testing.T) {
+	database := "testdropapplicationdbs"
+
+	couchInstance, err := createCouchInstance(config, &disabled.Provider{})
+	assert.NoError(t, err, "Error when trying to create couch instance")
```
I understand that this whole test file follows a similar assertion statement with an error message. I don't know how this will help us. `assert.NoError(t, err)` is a much cleaner way and more readable. We get the error only when we add new features or change existing code. At that time, this error message does not help much in debugging. We look at the line number and directly go to the test file to see why the error has occurred. We also refer to debug logs. For this PR, it is fine as existing UT func also uses similar statements.  Going forward, we can avoid using an error message within the assertion when it is not necessary as it affects the readability (need to read some extra words which does add much value).

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil_test.go:
```diff
@@ -184,3 +186,35 @@ func TestConstructedNamespaceDBName(t *testing.T) {
+
+	dbs, err := couchInstance.retrieveApplicationDBNames()
+	assert.NoError(t, err, "Error when retrieving application db names")
+	assert.Equal(t, 1+numCouchdbs, len(dbs), "Number of databases created")
```
the error message seems incomplete.

cendhu commented on core/ledger/kvledger/upgrade_dbs.go:
```diff
@@ -24,29 +32,39 @@ func UpgradeDBs(rootFSPath string) error {
+	if bytes.Equal(format, []byte(dataformat.Version20)) {
+		logger.Info("Ledger data format is current, do nothing.")
+		return nil
+	}
```
This is checked twice. Once here and again at line 59. This is because `upgradeFormat()` has side-effect. As per the name of the function, it should upgrade the format. However, it is also checking whether the store has the latest format (which has been checked already here). I would suggest splitting `idStore.upgradeFormat()` into `idStore.hasLastestFormat()` and `idStore.upgradeFormat()`. We can also use `idStore.hasVersion20Format()` to be explicit. Then, replace these lines with the newly introduced function.  We can also replace a few lines in `openIDStore()` with the newly introduced function but it is not related to this PR. 

With this change, we can get rid of the newly introduced function `idStore.getFormat()`.

wenjianqiao replied:
I recall that we purposely return an error if the idStore format is neither nil (1.x) or 2.0 with a forward thinking. When 3.0 if available, the method will only upgrade from 2.0 to 3.0 and return an error for other formats. Given it, idStore.hasLastestFormat() may not be helpful if there are more than 2 formats. Of course, we can simplify the code for now since we only have 1.x and 2.0.
@manish-sethi what do you think?

cendhu replied:
I am not questioning the return of an `error`. I mentioned the side-effect in upgradeFormat and lines 41 to 48 occur twice overall. The `updateFormat()` is in fact doing `updateFormatIfCurrentFormatIsNotV20()`. If we can split that into two, we can avoid this code duplication, execution duplication as well as the side-effect.

cendhu commented on core/ledger/kvledger/upgrade_dbs.go:
```diff
@@ -24,29 +32,39 @@ func UpgradeDBs(rootFSPath string) error {
-	// upgrades idStore (ledgerProvider) and drop databases
-	if err := UpgradeIDStoreFormat(rootFSPath); err != nil {
+	idStore := &idStore{db, dbPath}
+	format, err := idStore.getFormat()
```
before checking for the format, shouldn't we check whether the `idStore` is empty and return early. It is a remote case though.

wenjianqiao replied:
Good point. This is like a user error. idStore cannot be empty unless something is wrong, e.g., ledger is corrupted or fileSystemPath in core.yaml points to a wrong path.

cendhu commented on core/ledger/kvledger/upgrade_dbs_test.go:
```diff
@@ -62,22 +54,104 @@ func TestUpgradeDBs(t *testing.T) {
 }
 
-func TestUpgradeIDStoreWrongFormat(t *testing.T) {
+func TestUpgradeDBsWithV13CouchDB(t *testing.T) {
```
For rollback and reset commands, we have added tests in the ledger-level integration test pkg. This is because these commands modify blockstore and many DBs (affects multiple sub-packages in the ledger). The upgradeDB command is also modifying blockstore and manyDBs. Further, `TestUpgradeDBsWithV13CouchDB()` and `TestUpgradeDBsWithV13CouchDB()` access testdata present in the ledger-level integration test pkg. Hence, I would prefer moving these tests to `kvledger/tests` to maintain consistency. 

This reminds me of the discussion @manish-sethi and I had a year back on moving recovery-related tests to `kvledger/tests` pkg as it involves multiple sub-packages in the ledger.

wenjianqiao replied:
Thank you, Senthil. Agree that we should have ledger-level integration tests for UpgradeDBs under kvledger/tests. However, these tests are meant for unit tests to verify idStore is updated correctly and databases are dropped or not dropped as expected.

cendhu replied:
In `kvledger/tests`, we do have [verifyRebuilableDoesNotExist()](https://github.com/hyperledger/fabric/blob/master/core/ledger/kvledger/tests/env.go#L144) and [verifyRebuilablesExist()](https://github.com/hyperledger/fabric/blob/master/core/ledger/kvledger/tests/env.go#L126) which checks whether the DBs has been dropped or DBs exist. We use these in both reset and rollback tests. These functions can be extended to add a CouchDB check too. You can find no UT for rollback and reset commands in `kvledger` package. 

As I mentioned in one of the above comments, if we have `hasVersion20Format()`, we can also add an exported function to `test_export.go` and use it in the `kvledger/tests` to check whether the format has been applied.

We can discuss further and can be done in a separate PR too.

wenjianqiao commented:
@cendhu thank you for your comments. I have addressed most of your comments except that I didn't rename idStore getFormat to hasVersion20Format() because I want to avoid adding such a method for each new format.

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil_test.go:
```diff
@@ -184,3 +186,35 @@ func TestConstructedNamespaceDBName(t *testing.T) {
+
+	dbs, err := couchInstance.retrieveApplicationDBNames()
+	assert.NoError(t, err, "Error when retrieving application db names")
+	assert.Equal(t, 1+numCouchdbs, len(dbs), "Expected number of databases are created")
```
```suggestion
	assert.Equal(t, 1+numCouchdbs, len(dbs), "Expected number of databases are not created")
```

cendhu commented on core/ledger/kvledger/upgrade_dbs.go:
```diff
@@ -24,29 +33,44 @@ func UpgradeDBs(rootFSPath string) error {
-	if err := UpgradeIDStoreFormat(rootFSPath); err != nil {
-		return err
+	if emptydb, _ := db.IsEmpty(); emptydb {
+		return errors.Errorf("ledger database %s is empty. Make sure core.yaml has the correct ledger path.", dbPath)
```
it will be empty when the peer has joined no channel. Anyway, as I mentioned earlier, it is a remote corner case.

manish-sethi replied:
I don't have a strong feeling either way to keep or remove this check. However, to take a note, another problem with this check (extending the no channel use case) is that it will change behavior if you start the peer once and then execute the upgrade command again. Because, then the version entry would have been added and this condition won't be true. 

From the user point of view, even then there won't be any data present  but this message will no longer be shown.

wenjianqiao replied:
I keep the check because otherwise we cannot tell why formatKey does not exist (either 1.x version or empty db).

cendhu reviewed (commented):
Thank you, @wenjianqiao,  for the changes. Overall, LGTM. I yet to look at tests in detail. 

> I didn't rename idStore getFormat to hasVersion20Format()

This is completely fine. My major concern was that the following code was duplicated and executed twice (+ some side-effect). In the latest commit, you have fixed this issue. We can even remove the `getFormat()` as it used at only one place and move `s.db.Get()` to the place where it calls `getFormat()`. Don't need to push anything until Manish reviews this PR. 

```
        format, err := s.db.Get(formatKey)
	if err != nil {
		return err
	}
	idStoreFormatBytes := []byte(dataformat.Version20)
	if bytes.Equal(format, idStoreFormatBytes) {
		logger.Debug("Format is current, nothing to do")
		return nil
	}
	if format != nil {
		err = &dataformat.ErrVersionMismatch{
			ExpectedVersion: "",
			Version:         string(format),
			DBInfo:          fmt.Sprintf("leveldb for channel-IDs at [%s]", s.dbPath),
		}
		return err
	}
```

manish-sethi commented on core/ledger/kvledger/upgrade_dbs.go:
```diff
@@ -24,29 +33,44 @@ func UpgradeDBs(rootFSPath string) error {
+			ExpectedVersion: "",
+			Version:         string(format),
+			DBInfo:          fmt.Sprintf("leveldb for channel-IDs at [%s]", dbPath),
+		}
```
We had this check deliberately  in the function `idStore.upgradeFormat`. So that we do not accidentally end-up using that function without a precondition check somewhere.  I suggest to keep it there for sure. I am fine, if you want to check here again.

Also, like `dataformat.Version20`, it's better to use `dataformat.Version1x` explicitly instead of `nil`, or `""`. For byte comparison, you may have to define `Version1xBytes` explicitly though.

Not for this PR but I think that we discussed previously that we can introduce two generic constants `CurrentFormat` and `PreviousFormat` instead of `Version20` and `Version1x`.

cendhu replied:
> We had this check deliberately in the function idStore.upgradeFormat. So that we do not accidentally end-up using that function without a precondition check somewhere. I suggest to keep it there for sure. I am fine, if you want to check here again.

I suggested this to be removed from `idStore.upgradeFormat()` because it introduces code duplication, execution duplication, and side-effect in the function `upgradeFormat()`. Currently, `idStore.upgradeFormat()` is used only by the `UpgradeDBs()`. I don't see a strong justification for code duplication, execution duplication, and side-effect but I will step aside if you see it.

manish-sethi replied:
I gave the justification in my comment - which is code robustness. A function better not be knowing that how it is being used and hence preconditions be checked inside. 
There are situations when we cannnot avoid that e.g., assume that the lock has been acquired by the caller. But there you still prefer to keep function name either xxxwithoutLock or specify in the comments.
Here, repeating an if check twice in the code flow is not such a big problem if its helping in early return as well as code strength.

What side-effect are you referring to?

cendhu replied:
For the above-mentioned justification only, I mentioned it is not strong enough (for me). At the same time, if you are okay with it, I have no problem in having the following.  

- at least 10 lines are duplicated.
- in the UpgradeDBs() execution flow, these 10 lines occurs twice.
- only one functions call updateFormat. Hence, I don't know about the robustness. I also couldn't speculate what other future functions would use it and what are the probabilities that we would miss checking the current format before calling updateFormat. 
- `updateFormat()` is more like `updateFormatIfNotCurrent()`

wenjianqiao replied:
I moved the checking per review comment mainly because it is a package-only method - we own the caller code to do the checking before calling. However, I agree with Manish to also keep the checking in idStore.upgradeFormat. 
I propose to add a func idStore.checkFormatUpgradeEligibility that will be called by UpgradeDBs and idStore.ugradeFormat. This way idStore.getFormat is not needed any more.

cendhu replied:
As the majority feels the other way, I will step aside. Sorry, @wenjianqiao for the extra work incase you have to revert the changes.

manish-sethi commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdbutil_test.go:
```diff
@@ -184,3 +186,35 @@ func TestConstructedNamespaceDBName(t *testing.T) {
+	assert.NoError(t, err, "Error when trying to create couch instance")
+
+	err = writeDataFormatVersion(couchInstance, dataformat.Version20)
+	assert.NoErrorf(t, err, "Error when trying to write data format version to %s", fabricInternalDBName)
```
what is the significance of writing this explicitly for this test?

wenjianqiao replied:
It was called to create the fabric__internal database. Changed to call createCouchDatabase(couchInstance, fabricInternalDBName) instead.

manish-sethi commented on core/ledger/kvledger/tests/v1x_test.go:
```diff
@@ -107,18 +82,45 @@ func TestV13WithStateCouchdb(t *testing.T) {
 
 	// drop couchDB and other rebuildable dbs to test db rebuild
-	dropCouchDBs(t, couchdbConfig)
+	require.NoError(t, statecouchdb.DropApplicationDBs(couchdbConfig))
```
Not in this PR but just noticed - though we are not writing anything in parallel here, still better to move this line after the next line (i.e., after closing the ledger).

manish-sethi reviewed (commented):
LGTM except a couple minor comments. 
The test files in kvledger/tests are more readable now, by splitting the tests and moving couch data loading in a separate function.

wenjianqiao commented:
@manish-sethi @cendhu 
Thank you for your comments. I have updated code to address your comments. The checking for empty db is kept because otherwise we cannot tell why formatKey does not exist (either 1.x version or empty db). Added a new checkUpgradeEligibility func that is called by both UpgradeDBs and upgradeFormat. Removed getFormat.
@manish-sethi  for PreviousVersion and CurrentVersion change, plan to use a separate PR since it impacts other dbs.

manish-sethi commented on core/ledger/kvledger/kv_ledger_provider.go:
```diff
@@ -498,30 +498,51 @@ func openIDStore(path string) (s *idStore, e error) {
-	if format != nil {
+	if !bytes.Equal(format, dataformat.Version1xBytes) {
 		err = &dataformat.ErrVersionMismatch{
 			ExpectedVersion: "",
```
Missed using the constant here but do not want to hold this PR for this.
